explorer/testdata/linked_list/typed_linked_list.carbon

```diff
@@ -0,0 +1,139 @@
+    return {
+      .head = Optional(Node(T)*).CreateEmpty(),    
+      .tail = Optional(Node(T)*).CreateEmpty(),
+      .len = 0
+    };
+  }
+  fn PushBack[addr me: Self*](value:T) {
+    (*me).len = (*me).len + 1;
+    if(not (*me).head.HasValue()) {
+      (*me).head = Optional(Node(T)*).Create(heap.New(Node(T).Create(value)));
+      (*me).tail = (*me).head;
+      return;
+    }
+    var last: Optional(Node(T)*) = (*me).tail;
+    var last_value:Node(T)* = last.Get(); 
+    var v_wrapped:Optional(Node(T)*) = Optional(Node(T)*).Create(heap.New(Node(T).Create(value)));
+    (*last_value).set_next(v_wrapped);
+    (*v_wrapped.Get()).set_prev(last);
+    (*me).tail = v_wrapped;
+  }
+  fn PushFront[addr me: Self*](value:T) {
+    (*me).len = (*me).len + 1;
+    if(not (*me).head.HasValue()) {
+      (*me).head = Optional(Node(T)*).Create(heap.New(Node(T).Create(value)));
+      (*me).tail = (*me).head;
+      return;
+    }
+    var v_wrapped:Optional(Node(T)*) = Optional(Node(T)*).Create(heap.New(Node(T).Create(value)));
+    var current_head: Optional(Node(T)*) = (*me).head;
+    var current_head_value: Node(T)* = current_head.Get();
+    (*v_wrapped.Get()).set_next(current_head);
+    (*current_head_value).set_prev(v_wrapped);
+    (*me).head = v_wrapped;
+  }
+  fn GetByIndex[me: Self](target_index: i32) -> Optional(T) {
+    if(target_index > me.len - 1 or target_index < 0) {
+      // Not in possible range
+      return Optional(T).CreateEmpty();
+    }
+    var search_backwards:bool = target_index > me.len / 2;
+    var iter:Optional(Node(T)*) = if search_backwards then me.tail else me.head;
+    var c:i32 = if search_backwards then me.len -1 else 0;
+    while(c != target_index) {
+      var node: Node(T) = *(iter.Get());
+      if(search_backwards) {
+        c = c - 1;
+        iter = node.prev;
+      } else {
+        iter = node.next;
+        c = c + 1;
+      }
+    }
+    return (*(iter.Get())).value;
+  }
+  fn Clear[addr me: Self*]() {
+    if((*me).len == 0) {
+      return;
+    } 
+    var iter: auto = (*me).head;
+    while(iter.HasValue()) {
+      var current: auto = iter;
+      iter = (*iter.Get()).next;
+      heap.Delete(current.Get());
+    }
+    (*me).head = Optional(Node(T)*).CreateEmpty();
+    (*me).tail = Optional(Node(T)*).CreateEmpty();
+    (*me).len = 0; 
+  }
+  fn Length[me:Self]() -> i32 {
+    return me.len;
+  }
+  var head: Optional(Node(T)*);
+  var tail: Optional(Node(T)*);
+  var len: i32;
+}
+
+fn Main() -> i32 {
+  var test_list:auto = LinkedList(String).Create();
+  test_list.PushBack("This is a String Entry 0");
+  test_list.PushBack("This is a String Entry 1");
+  test_list.PushBack("This is a String Entry 2");
+  test_list.PushBack("This is a String Entry 3");
+  test_list.PushBack("This is a String Entry 4");
+  test_list.PushBack("This is a String Entry 5");
+  test_list.PushFront("This is a prepended String -1");
+  Print("String array length is: {0}", test_list.Length());
```

Comment:
I think this should say "list" rather than "array"?

Reply:
👍